### PR TITLE
Add option to disable quantizing specified node kinds

### PR DESCRIPTION
**Pull request by jfix71**
*Description*: We may prefer to not quantize every node kind in a model if it causes unacceptable accuracy loss. This commit adds the ability to disable quantization for node kinds that would otherwise be quantized, preventing accuracy loss.

*Testing*: Added a unit test.

*Documentation*: Added.

### CODE REVIEW

File: include/glow/Quantization/Quantization.h [hunk 1]
```diff
@@ -62,13 +62,16 @@ generateNodeQuantizationInfos(const Function *F,
 /// function based on \p quantizationInfos. This method converts to integer as
 /// many nodes as permitted by the backend \p EE. The new quantized function is
 /// called \p newFuncName. If no name is given the method will generate a name.
-/// This method clones original function \p F and caller is responsible
-/// for cleaning up/erasing original function \p F if needed.
+/// This method clones original function \p F and caller is responsible for
+/// cleaning up/erasing original function \p F if needed. Any nodes of kinds
+/// contained in \p doNotQuantizeKinds will not be quantized, even if a profile
+/// was gathered for them.
```
Comment (rdzhabarov) [inline include/glow/Quantization/Quantization.h]: `if a profile was gathered for them and backend supports quantized operation`?

File: docs/Quantization.md [hunk 2]
```diff
@@ -102,6 +102,17 @@ the graph.
...
+By default, all nodes that can be quantized will be quantized. However, we may
+only want to quantize some parts of a model, e.g. if accuracy loss is too high
+when all node kinds are quantized. The Glow loader currently allows for
+disabling quantization of all nodes of a specific kind which are found in the
+graph. For example, if the loaded model sees high accuracy loss when
+element-wise Add is quantized, it can be left in floating point. This can be
+done by passing on the command line the node name via the option
+`-do_not_quantize_node`. For example, to not quantize the Add node when running
+the quantized text translator: ``` ./bin/text-translator -m en2gr
+-load_profile=en2gr.yaml -do_not_quantize_node=Add ```
```
Comment (rdzhabarov) [inline docs/Quantization.md]: can you mention that `do_not_quantize_node` could contain several node types? I got a sense that it's only a single node type can be blacklisted here.

File: tests/unittests/quantizationTest.cpp [hunk 3]
```diff
@@ -659,6 +659,78 @@ TEST(Quantization, quantizeSoftmaxAndLRN) {
...
+
+  std::vector<NodeQuantizationInfo> QI{
+      {NodeQuantizationInfo::generateNodeOutputName(LHS->getName()), {0.3f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(RHS->getName()), {0.4f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(MMN->getName()), {0.6f, 0}},
+      {NodeQuantizationInfo::generateNodeOutputName(TN->getName()), {0.5f, 0}},
+  };
+
+  // Do not quantize the tanh.
+  KindSet doNotQuantize;
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: does this not work: KindSet doNotQuantize = {Kinded::Kind::TanhNodeKind} ?

File: docs/Quantization.md (hunk 2 above)
Comment (lygstate) [reply docs/Quantization.md]: Is there a more complete way to do that?
Such as an sub-group of nodes be disabled for quantize?

File: tests/unittests/quantizationTest.cpp (hunk 3 above)
Comment (jfix71) [reply tests/unittests/quantizationTest.cpp]: Unfortunately not 😞

File: docs/Quantization.md (hunk 2 above)
Comment (jfix71) [reply docs/Quantization.md]: Sure -- just to be clear I believe the way to add multiple nodes here is to specify each one individually, e.g. `-do_not_quantize_node=Add -do_not_quantize_node=Div` etc. Hence why I made it node and not nodes. Is that not the case, is there a way to do all under one flag?

Comment (jfix71) [reply docs/Quantization.md]: @lygstate Are you talking about specifying multiple Node kinds, as I replied above? Or by a sub-group of nodes do you mean that only some of the nodes with the Node kind would be quantized?

File: tests/unittests/quantizationTest.cpp [hunk 4]
```diff
@@ -659,6 +659,78 @@ TEST(Quantization, quantizeSoftmaxAndLRN) {
...
+  doNotQuantize.insert(Kinded::Kind::TanhNodeKind);
```
[the remark this answers is not in the document]
Comment (jfix71) [reply tests/unittests/quantizationTest.cpp]: Done

File: include/glow/Quantization/Quantization.h (hunk 1 above)
Comment (jfix71) [reply include/glow/Quantization/Quantization.h]: Done

File: docs/Quantization.md (hunk 2 above)
Comment (qcolombet) [reply docs/Quantization.md]: Using comma should work as well, IIRC the LLVM option parsing:
 `-do_not_quantize_node=Add,Div`

Comment (jfix71) [reply docs/Quantization.md]: Figured it out -- I needed to add `llvm::cl::CommaSeparated` to enable that. Thanks @qcolombet!

**Pull request merged**